synapse/tests/test_cortex.py

```diff
@@ -988,6 +988,100 @@ def runtufobydefault(self, core):
+            ('splice', {'mesg': ('zillion', {})}),
+
+            ('splice', {'mesg': ('not-in-the-rules', {'hehe': 'haha'})}),
+        )
+        expected = [
+            ('node:add', {'form': 'thing', 'valu': 'newp'}),
+            ('node:add', {}),
+            ('node:del', {'form': 'thing'}),
+            ('yeap', {}),
+        ]
+
+        def run_tests(core, msgs, expected):
+            [core.dist(msg) for msg in msgs]
+
+            # subscribe to the membrane fifo and check the messages that made it through
+            actual = []
+            core.subCoreFifo(name, actual.append)
+
+            actual_msgs = [msg[2] for msg in actual]
+            self.eq(actual_msgs, expected)
+
+        # Add a membrane via Cortex API
+        with self.getTestDir() as dirn:
+            with s_cortex.openurl('dir:///' + dirn) as core:
+                core.addCoreMembrane(name, rules)
+                run_tests(core, msgs, expected)
+
+        # Add membranes via setting config opt
+        with self.getTestDir() as dirn:
+            with s_cortex.openurl('dir:///' + dirn) as core:
+                core.setConfOpt('membranes', ((name, rules),))
+                run_tests(core, msgs, expected)
+
+        # Test missing and dup membranes
+        with self.getTestDir() as dirn:
+            with s_cortex.openurl('dir:///' + dirn) as core:
+
+                self.raises(NoSuchMembrane, core.delCoreMembrane, name)
+                core.addCoreMembrane(name, rules)
+
+                [core.dist(msg) for msg in msgs]
+                actual = []
+                core.subCoreFifo(name, actual.append)
+                actual_msgs = [msg[2] for msg in actual]
+                self.eq(actual_msgs, expected)
+
+                self.raises(MembraneExists, core.addCoreMembrane, name, rules)
+                actual = []
+                core.subCoreFifo(name, actual.append)
+                actual_msgs = [msg[2] for msg in actual]
+                self.eq(actual_msgs, expected)
+
+                # remove and re-add the membrane, subscribe again and make sure it is empty
+                fifo_dir = core.getCoreFifo(name).getConfOpt('dir')
+                self.gt(len(os.listdir(fifo_dir)), 0)  # did should be populated
+                self.none(core.delCoreMembrane(name))  # returns none
+                self.raises(FileNotFoundError, os.listdir, fifo_dir)  # should not exist
+                self.len(0, core.getTufosByProp('syn:fifo'))
+                self.raises(NoSuchFifo, core.getCoreFifo, name)  # should be removed
+                self.raises(NoSuchFifo, core.subCoreFifo, name, actual.append)
+                self.raises(NoSuchMembrane, core.delCoreMembrane, name)
+
+                core.addCoreMembrane(name, rules)
+                actual = []
+                core.subCoreFifo(name, actual.append)
+                actual_msgs = [msg[2] for msg in actual]
+                self.eq(actual_msgs, [])  # There should be no messages in the membrane yet
```

Comment:
FIXME - this test still fails. Somehow the messages are present in the fifo before we add them.

Reply:
Fixed

Reply:
Fixed